azure-iot-device/tests/common/pipeline/test_pipeline_stages_base.py

```diff
@@ -107,12 +99,441 @@ def mock_execute_op(self, op):
+        connect_op.error = fake_exception
+        operation_flow.complete_op(stage=stage.next, op=connect_op)
+
+        assert_callback_failed(op=op, error=fake_exception)
+
+    @pytest.mark.it("Waits for the ConnectOperation to complete before pasing the operation down")
+    def test_connect_success(self, params, op, stage):
+        stage.pipeline_root.connected = False
+
+        stage.run_op(op)
+        assert stage.next.run_op.call_count == 1
+        connect_op = stage.next.run_op.call_args[0][0]
+        operation_flow.complete_op(stage=stage.next, op=connect_op)
+
+        assert stage.next.run_op.call_count == 2
+        assert stage.next.run_op.call_args[0][0] == op
+
+    @pytest.mark.it("calls the op's callback when the operation is complete after connecting")
+    def test_operation_complete(self, params, op, stage):
+        stage.pipeline_root.connected = False
+
+        stage.run_op(op)
+        connect_op = stage.next.run_op.call_args[0][0]
+        operation_flow.complete_op(stage=stage.next, op=connect_op)
+
+        operation_flow.complete_op(stage=stage.next, op=op)
+        assert_callback_succeeded(op=op)
+
+    @pytest.mark.it("calls the op's callback when the operation fails after connecting")
+    def test_operation_fails(self, params, op, stage, fake_exception):
+        stage.pipeline_root.connected = False
+
+        stage.run_op(op)
+        connect_op = stage.next.run_op.call_args[0][0]
+        operation_flow.complete_op(stage=stage.next, op=connect_op)
+        op.error = fake_exception
+        operation_flow.complete_op(stage=stage.next, op=op)
+
+        assert_callback_failed(op=op, error=fake_exception)
+
+
+pipeline_stage_test.add_base_pipeline_stage_tests(
+    cls=pipeline_stages_base.SerializeConnectOpsStage,
+    module=this_module,
+    all_ops=all_common_ops,
+    handled_ops=[
+        pipeline_ops_base.ConnectOperation,
+        pipeline_ops_base.DisconnectOperation,
+        pipeline_ops_base.ReconnectOperation,
+    ],
+    all_events=all_common_events,
+    handled_events=[],
+    extra_initializer_defaults={"blocked": False, "queue": queue.Queue},
+)
+
+connect_ops = [
+    {"op_class": pipeline_ops_base.ConnectOperation, "connected_flag_required_to_run": False},
+    {"op_class": pipeline_ops_base.DisconnectOperation, "connected_flag_required_to_run": True},
+    {"op_class": pipeline_ops_base.ReconnectOperation, "connected_flag_required_to_run": True},
+]
+
+
+class FakeOperation(pipeline_ops_base.PipelineOperation):
+    pass
+
+
+@pytest.mark.describe("SerializeConnectOpsStage - .run_op() -- called with connect operation")
+class TestSerializeConnectOpStageRunOp(object):
+    @pytest.fixture
+    def stage(self, mocker):
+        stage = make_mock_stage(
+            mocker=mocker, stage_to_make=pipeline_stages_base.SerializeConnectOpsStage
+        )
+        stage.next.run_op = mocker.MagicMock()
+        return stage
+
+    @pytest.fixture
+    def connect_op(self, mocker, params):
```

Comment:
this should probably also be `connection_op` rather than `connect_op` since it could apply to not just `ConnectOperation`, but also `DisconnectOperation` and `ReconnectOperation`.

although admittedly, that terminology isn't much better (especially if we end up eventually turning the transport into something called a "connection") as per the Track 2 stuff.